python/tank/descriptor/io_descriptor/git.py

```diff
@@ -43,17 +56,109 @@ def __init__(self, descriptor_dict):
         # Note: the git command always uses forward slashes
         self._sanitized_repo_path = self._path.replace(os.path.sep, "/")
 
-    def _clone_repo(self, target_path):
+    @LogManager.log_timing
+    def _clone_then_execute_git_command(self, target, commands):
         """
-        Clone the repo into the target path
+        Clones the git repository into the given location and
+        executes the given list of git commands.
+
+        The initial clone operation happens via an os.system call, ensuring
+        that there is an initialized shell environment, allowing git
+        to potentially request shell based authentication for repositories
+        which requires credentials.
 
-        :param target_path: The target path to clone the repo to
-        :raises:            TankError if the clone command fails
+        The subsequent list of commands are intended to be executed on the
+        recently cloned repository and will the cwd will be set so that they
+        are executed in the directory scope of the newly cloned repository.
+
+        :param target: path to clone into
+        :param commands: list git commands to execute, e.g. ['checkout x']
+        :returns: stdout and stderr of the last command executed as a string
+        :raises: TankGitError on git failure
         """
+        # ensure *parent* folder exists
+        parent_folder = os.path.dirname(target)
+        filesystem.ensure_folder_exists(parent_folder)
+
+        # first probe to check that git exists in our PATH
+        log.debug("Checking that git exists and can be executed...")
+        try:
+            output = subprocess_check_output(["git", "--version"])
+        except:
+            raise TankGitError(
+                "Cannot execute the 'git' command. Please make sure that git is "
+                "installed on your system and that the git executable has been added to the PATH."
+            )
+        log.debug("Git installed: %s" % output)
+
         # Note: git doesn't like paths in single quotes when running on
-        # windows - it also prefers to use forward slashes!
-        log.debug("Git Cloning %r into %s" % (self, target_path))
-        execute_git_command("clone -q \"%s\" \"%s\"" % (self._sanitized_repo_path, target_path))
+        # windows - it also prefers to use forward slashes
+        #
+        # note that we use os.system here to allow for git to pop up (in a terminal
+        # if necessary) authentication prompting. This DOES NOT seem to be possible
+        # with subprocess.
+        #
+        log.debug("Git Cloning %r into %s" % (self, target))
+        cmd = "git clone -q \"%s\" \"%s\"" % (self._path, target)
```

Comment:
it's a windows thing (see comments above). I'll move the comments to be a bit closer to the command.